odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return self.vector * self.operator.gradient * self.vector
+
+    # TODO: can this be computed?
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+
+        Notes
+        -----
+        The nonsmooth solvers that make use of proximal operators in order to
+        solve a given optimization problem, see for example
+        `forward_backward_pd`, take a `proximal factory` as input. Note that
+        ``Functional.proximal`` is in fact a `proximal factory`.
+        """
+        raise NotImplementedError('there is no known expression for this')
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of the functional.
+
+        Notes
+        -----
+        The convex conjugate functional of a convex functional :math:`f(x)`,
+        defined on a Hilber space, is defined as the functional
+
+        .. math::
+
+            f^*(x^*) = \\sup_{x} \{ \\langle x^*,x \\rangle - f(x)  \}.
+
+        The concept is also known as the Legendre transformation.
+
+        References
+        ----------
+        Wikipedia article on `Convex conjugate
+        <https://en.wikipedia.org/wiki/Convex_conjugate>`_.
+
+        Wikipedia article on `Legendre transformation
+        https://en.wikipedia.org/wiki/Legendre_transformation
+
+        For literature references see, e.g., [Lue1969]_, [Roc1970]_.
+        """
+        return self.operator.conjugate_functional * (1.0 / self.vector)
+
+
+class FunctionalSum(Functional, OperatorSum):
+
+    """Expression type for the sum of functionals.
+
+    ``FunctionalSum(func1, func2) <==> (x --> func1(x) + func2(x))``.
+    """
+
+    def __init__(self, func1, func2, tmp_dom=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func1, func2 : `Functional`
+            The summands of the functional sum. Their `Operator.domain`
+            and `Operator.range` must coincide.
+        tmp_dom : `Operator.domain` `element`, optional
+            Used to avoid the creation of a temporary when applying the
+            gradient.
+        """
+        # TODO: Tmp_dom, what is it used for?
+        if not isinstance(func1, Functional):
+            raise TypeError('functional 1 {!r} is not a Functional instance.'
```

Comment:
```
`Functional`
```
in error messages, also 
```
`func1`
```
(explicit parameter names). Change across the board.